rs/execution_environment/src/scheduler.rs

```diff
@@ -2376,3 +2390,49 @@ fn scheduled_heap_delta_limit(
         .saturating_sub(remaining_heap_delta_reserve)
         .into()
 }
+
+/// Aggregator and observer of per-canister ingress queue latencies.
+struct CanisterIngressQueueLatencies {
+    /// Per canister observed ingress message latency sum and count.
+    latencies: BTreeMap<PrincipalId, (f64, usize)>,
+    /// Current block time.
+    time: Time,
+    /// Histogram to observe the latencies.
+    histogram: Histogram,
+}
+
+impl CanisterIngressQueueLatencies {
+    fn new(time: Time, histogram: Histogram) -> Self {
+        Self {
+            latencies: BTreeMap::new(),
+            time,
+            histogram,
+        }
+    }
+
+    /// Records the ingress queue latency of a message iff it is transitioning from
+    /// `Received` to some other state (i.e. when popped from the ingress queue).
+    fn on_ingress_status_changed(&mut self, old_status: Arc<IngressStatus>) {
+        if let IngressStatus::Known {
+            receiver,
+            user_id: _,
+            time,
+            state: IngressState::Received,
+        } = &*old_status
+        {
+            let (latency, count) = self.latencies.entry(*receiver).or_default();
+            *latency += self.time.saturating_duration_since(*time).as_secs_f64();
+            *count += 1;
+        }
+    }
+}
+
+impl Drop for CanisterIngressQueueLatencies {
+    /// Observes the average ingress queue latency of each canister at the end of
+    /// the round.
+    fn drop(&mut self) {
+        for (latency, count) in self.latencies.values() {
+            self.histogram.observe(*latency / *count as f64);
+        }
+    }
```

Comment:
Just a general thing I wonder sometimes here and also some other places. When working with experimental data one cares as much about the standard deviation as the average. Wouldn't it make sense to record that as well in case we get complaints about flaky response to ingress messages or something? Or rather for us to notice that before we get complaints.

Reply:
> And most of the time it would be exceedingly boring.
Well yes but for the times where it gets interesting it really would be, no?

I mean it would probably mean a fair bit of code unless it's implemented already so I'm perfectly fine the way it is. It's just that the spread is often just as interesting or sometimes even more interesting than the mean (generally speaking).

But IIUC what you're saying the source of more noise there would be the canisters themselves? In that case I guess it wouldn't be particularly useful here I agree.

Reply:
Feel free to merge it as it is.